File: 43.multiply-strings.py
#
# @lc app=leetcode id=43 lang=python3
#
# [43] Multiply Strings
#

# Time:  O(m * n)
# Space: O(m + n)

# @lc code=start
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        len1, len2 = len(num1), len(num2)
        res = [0] * (len1 + len2)
        for i in range(len1 - 1, -1, -1):
            for j in range(len2 - 1, -1, -1):
                res[i + j + 1] += int(num1[i]) * int(num2[j])
                res[i + j] += res[i + j + 1] // 10
                res[i + j + 1] %= 10
        for i in range(len(res)):
            if res[i]:
                break
        return "".join(map(lambda x: str(x), res[i:]))


# @lc code=end

File: 43.multiply-strings.js
/*
 * @lc app=leetcode id=43 lang=javascript
 *
 * [43] Multiply Strings
 */

// @lc code=start
/**
 * @param {string} num1
 * @param {string} num2
 * @return {string}
 */
var multiply = function (num1, num2) {
  const len1 = num1.length;
  const len2 = num2.length;
  const res = Array(len1 + len2).fill(0);
  for (let i = len1 - 1; i >= 0; --i) {
    for (let j = len2 - 1; j >= 0; --j) {
      res[i + j + 1] += Number(num1[i]) * Number(num2[j]);
      res[i + j] += Math.trunc(res[i + j + 1] / 10);
      res[i + j + 1] %= 10;
    }
  }
  for (let i = 0; i < res.length; ++i) {
    if (res[i]) {
      return res.slice(i).join("");
    }
  }
  return "0";
};
// @lc code=end
